src/dash-table/derived/cell/dropdowns.ts

```diff
@@ -1,94 +1,128 @@
-};
+class Dropdowns {
+    /**
+     * Return the dropdown for each cell in the table.
+     */
+    get = memoizeOne((
+        columns: VisibleColumns,
+        data: Data,
+        indices: Indices,
+        columnConditionalDropdown: any,
+        columnStaticDropdown: any,
+        dropdown_properties: any
+    ) => mapData((datum, rowIndex) => R.map(column => {
+        const applicable = this.applicable.get(column.id, rowIndex)(
+            column,
+            indices[rowIndex],
+            columnConditionalDropdown,
+            columnStaticDropdown,
+            dropdown_properties
+        );
 
-const getter = (
-    astCache: (key: [ColumnId, number], query: string) => SyntaxTree,
-    columns: VisibleColumns,
-    data: Data,
-    indices: Indices,
-    columnConditionalDropdown: IConditionalColumnDropdown[],
-    columnStaticDropdown: IColumnDropdown[],
-    dropdown_properties: IDropdownProperties
-): (DropdownValues | undefined)[][] => mapData((datum, rowIndex) => R.map(column => {
-    const realIndex = indices[rowIndex];
+        return this.dropdown.get(column.id, rowIndex)(
+            applicable,
+            column,
+            datum
+        );
+    }, columns), data));
 
-    let legacyDropdown = (
-        (
-            dropdown_properties &&
-            dropdown_properties[column.id] &&
+    /**
+     * Returns the list of applicable dropdowns for a cell.
+     */
+    private readonly applicable = memoizerCache<[ColumnId, number]>()((
+        column: IBaseVisibleColumn,
+        realIndex: number,
+        columnConditionalDropdown: any,
+        columnStaticDropdown: any,
+        dropdown_properties: any
+    ): [any, any] => {
+        let legacyDropdown = (
             (
-                dropdown_properties[column.id].length > realIndex ?
-                    dropdown_properties[column.id][realIndex] :
-                    null
-            )
-        ) || column
-    ).options;
+                dropdown_properties &&
+                dropdown_properties[column.id] &&
+                (
+                    dropdown_properties[column.id].length > realIndex ?
+                        dropdown_properties[column.id][realIndex] :
+                        null
+                )
+            ) || column
+        ).options;
 
-    const conditional = columnConditionalDropdown.find((cs: any) => cs.id === column.id);
-    const base = columnStaticDropdown.find((ss: any) => ss.id === column.id);
+        const conditional = columnConditionalDropdown.find((cs: any) => cs.id === column.id);
+        const base = columnStaticDropdown.find((ss: any) => ss.id === column.id);
 
-    const conditionalDropdowns = (conditional && conditional.dropdowns) || [];
-    const staticDropdown = legacyDropdown || (base && base.dropdown);
+        return [
+            legacyDropdown || (base && base.dropdown),
+            (conditional && conditional.dropdowns) || []
+        ];
+    });
```

Comment:
Per cell caching -- this could probably be dropped but would require the use of a constant `[]` to not trigger re-renderers on React's side.